packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -33,70 +39,103 @@ function areDefinedAndEqual<T>(
   return (value1 || value2) !== undefined && value1 === value2;
 }
 
-function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
-  const results = [result];
-  if (result.parentResult) {
-    results.unshift(result.parentResult);
-  }
-  if (result.childResults?.length) {
-    results.push(
-      ...result.childResults.flatMap((childResult) =>
-        flattenResultTree(childResult)
-      )
-    );
+function getChildResultsRecursively(
+  result: ResultWithFolding
+): ResultWithFolding[] {
+  if (!result.childResults) {
+    return [];
   }
-  return results;
+  return result.childResults.flatMap((childResult) => [
+    childResult,
+    ...getChildResultsRecursively(childResult),
+  ]);
 }
 
-function foldResult(
-  source: ResultWithFolding,
+function resolveChildrenFromFields(
+  parent: ResultWithFolding,
   results: ResultWithFolding[],
   fields: FoldingFields
-): FoldedResult {
-  const sourceChildValue = getChildField(source, fields);
-  return {
-    ...source,
-    children: sourceChildValue
-      ? results
-          .filter((result) => {
-            const isSameResultAsSource =
-              getChildField(result, fields) === getChildField(source, fields);
-            const isChildOfSource =
-              getParentField(result, fields) === sourceChildValue;
-            return isChildOfSource && !isSameResultAsSource;
-          })
-          .map((result) => foldResult(result, results, fields))
-      : [],
-  };
+): FoldedResult[] {
+  const sourceChildValue = getChildField(parent, fields);
+  return sourceChildValue
```

Comment:
i'd do an early return here when `sourceChildValue` is undefined so it removes a bit of indentation and is easier to read